src/ray/raylet/node_manager.cc

```diff
@@ -1067,57 +1100,63 @@ void NodeManager::HandleWaitRequest(const rpc::WaitRequest &request,
     }
   }
 
-  const TaskID &current_task_id = TaskID::FromBinary(request.task_id());
+  const TaskID &current_task_id = from_flatbuf<TaskID>(*message->task_id());
   bool client_blocked = !required_object_ids.empty();
   if (client_blocked) {
-    HandleTaskBlocked(worker_id, required_object_ids, current_task_id, /*ray_get=*/false);
+    HandleTaskBlocked(client, required_object_ids, current_task_id, /*ray_get=*/false);
   }
 
   ray::Status status = object_manager_.Wait(
       object_ids, wait_ms, num_required_objects, wait_local,
-      [this, client_blocked, worker_id, current_task_id, reply, send_reply_callback](
-          std::vector<ObjectID> found, std::vector<ObjectID> remaining) {
-        IdVectorToProtobuf<ObjectID, rpc::WaitReply>(found, *reply,
-                                                     &rpc::WaitReply::add_found);
-        IdVectorToProtobuf<ObjectID, rpc::WaitReply>(remaining, *reply,
-                                                     &rpc::WaitReply::add_remaining);
-
-        // Send reply to finish this wait request.
-        send_reply_callback(Status::OK(), nullptr, nullptr);
-        if (client_blocked) {
-          HandleTaskUnblocked(worker_id, current_task_id);
+      [this, client_blocked, client, current_task_id](std::vector<ObjectID> found,
+                                                      std::vector<ObjectID> remaining) {
+        // Write the data.
+        flatbuffers::FlatBufferBuilder fbb;
+        flatbuffers::Offset<protocol::WaitReply> wait_reply = protocol::CreateWaitReply(
+            fbb, to_flatbuf(fbb, found), to_flatbuf(fbb, remaining));
+        fbb.Finish(wait_reply);
+
+        auto status =
+            client->WriteMessage(static_cast<int64_t>(protocol::MessageType::WaitReply),
```

Comment:
I guess it wasn't like this before, but while we're at it, do you want to make this a `WriteMessageAsync`?